File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = s => {
    const stack = [];
    const map = {
        ')': '(',
        ']': '[',
        '}': '{'
    }
    for (let i = 0; i < s.length; i++) {
        if ('([{'.includes(s[i])) stack.push(s[i]);
        else if (stack[stack.length - 1] === map[s[i]]) stack.pop();
        else stack.push(s[i]);
    }
    return stack.length === 0;
}

File: 0020-valid-parentheses/0020-valid-parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
    stack = []
    s.each_char do |char| 
        if char == ')' && stack[-1] == '('
            stack.pop()
        elsif char == '}' && stack[-1] == '{'
            stack.pop()
        elsif char == ']' && stack[-1] == '['
            stack.pop()
        else
            stack.push(char)
        end
    end
    return stack.length == 0
end